### ISSUE TITLE
Variational dropout in Augmented LSTM

### ISSUE THREAD

**Issue by apmoore1** [allenai/allennlp#2320]
**This is more a question and feature/bug I think**
From what I can tell from the [Augmented LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/augmented_lstm.py) the same variational dropout mask is applied to the recurrent state as well as the output of each LSTM cell. I don't think this is the same dropout principal that is stated in the [reference paper](https://arxiv.org/abs/1512.05287) where the dropout mask is different for the recurrent state and the output between each LSTM layer as shown in figure 1 of the paper. I was wondering if this is the case and if it would be possible to have two dropout masks one for the recurrent and one for the output just like the reference paper shows, of which I think there are two ways this can be solved:
1.  [Augmented LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/augmented_lstm.py) having two dropout arguments in it's constructor: `recurrent_dropout_probability` and `output_dropout_probability` and then having two seperate dropout masks one for recurrent and the other for output.
2. [Augmented LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/augmented_lstm.py) will only concern itself with the recurrent dropout and will not apply dropout to the output of each LSTM. The dropout between the LSTM layers will be resolved by adding `output_dropout_probability` to [Stacked Bi-Directional LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/stacked_bidirectional_lstm.py) and using [InputVariationalDropout](https://github.com/allenai/allennlp/blob/master/allennlp/modules/input_variational_dropout.py) on the output sequence of called LSTMs.

The benefit of doing point 2 in the list over 1 is that we can be more consistent with the [torch.nn.lstm](https://pytorch.org/docs/stable/nn.html#lstm) apporach to dropout where they only apply dropout to all output layers apart from the last. Where as if we go for point 1 in the list the dropout would be applied to all output layers including the last.

I am happy to implement this and create a pull request :smile:

**Comment by matt-gardner** [comment]
I would definitely do 2, to be consistent with our other encoders.  It's a problem if you have to change your code when you swap out a `Seq2SeqEncoders`.  PR welcome.
